// cashc/dist/main/ast/Globals.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GLOBAL_SYMBOL_TABLE = exports.Modifier = exports.Class = exports.TimeOp = exports.GlobalFunction = exports.NumberUnit = void 0;
const utils_1 = require("@cashscript/utils");
const SymbolTable_js_1 = require("./SymbolTable.js");
exports.NumberUnit = {
    SATOSHIS: 1,
    SATS: 1,
    FINNEY: 10,
    BITS: 100,
    BITCOIN: 100000000,
    SECONDS: 1,
    MINUTES: 60,
    HOURS: 3600,
    DAYS: 86400,
    WEEKS: 604800,
};
var GlobalFunction;
(function (GlobalFunction) {
    GlobalFunction["ABS"] = "abs";
    GlobalFunction["MIN"] = "min";
    GlobalFunction["MAX"] = "max";
    GlobalFunction["WITHIN"] = "within";
    GlobalFunction["RIPEMD160"] = "ripemd160";
    GlobalFunction["SHA1"] = "sha1";
    GlobalFunction["SHA256"] = "sha256";
    GlobalFunction["HASH160"] = "hash160";
    GlobalFunction["HASH256"] = "hash256";
    GlobalFunction["CHECKSIG"] = "checkSig";
    GlobalFunction["CHECKMULTISIG"] = "checkMultiSig";
    GlobalFunction["CHECKDATASIG"] = "checkDataSig";
})(GlobalFunction = exports.GlobalFunction || (exports.GlobalFunction = {}));
var TimeOp;
(function (TimeOp) {
    TimeOp["CHECK_SEQUENCE"] = "tx.age";
    TimeOp["CHECK_LOCKTIME"] = "tx.time";
})(TimeOp = exports.TimeOp || (exports.TimeOp = {}));
var Class;
(function (Class) {
    Class["LOCKING_BYTECODE_P2SH"] = "LockingBytecodeP2SH";
    Class["LOCKING_BYTECODE_P2PKH"] = "LockingBytecodeP2PKH";
    Class["LOCKING_BYTECODE_NULLDATA"] = "LockingBytecodeNullData";
})(Class = exports.Class || (exports.Class = {}));
var Modifier;
(function (Modifier) {
    Modifier["CONSTANT"] = "constant";
})(Modifier = exports.Modifier || (exports.Modifier = {}));
exports.GLOBAL_SYMBOL_TABLE = new SymbolTable_js_1.SymbolTable();
// Classes
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.class(Class.LOCKING_BYTECODE_P2SH, new utils_1.BytesType(23), [new utils_1.BytesType(20)]));
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.class(Class.LOCKING_BYTECODE_P2PKH, new utils_1.BytesType(25), [new utils_1.BytesType(20)]));
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.class(Class.LOCKING_BYTECODE_NULLDATA, new utils_1.BytesType(), [new utils_1.ArrayType(new utils_1.BytesType())]));
// Global functions
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.function(GlobalFunction.ABS, utils_1.PrimitiveType.INT, [utils_1.PrimitiveType.INT]));
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.function(GlobalFunction.MIN, utils_1.PrimitiveType.INT, [utils_1.PrimitiveType.INT, utils_1.PrimitiveType.INT]));
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.function(GlobalFunction.MAX, utils_1.PrimitiveType.INT, [utils_1.PrimitiveType.INT, utils_1.PrimitiveType.INT]));
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.function(GlobalFunction.WITHIN, utils_1.PrimitiveType.BOOL, [utils_1.PrimitiveType.INT, utils_1.PrimitiveType.INT, utils_1.PrimitiveType.INT]));
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.function(GlobalFunction.RIPEMD160, new utils_1.BytesType(20), [utils_1.PrimitiveType.ANY]));
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.function(GlobalFunction.SHA1, new utils_1.BytesType(20), [utils_1.PrimitiveType.ANY]));
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.function(GlobalFunction.SHA256, new utils_1.BytesType(32), [utils_1.PrimitiveType.ANY]));
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.function(GlobalFunction.HASH160, new utils_1.BytesType(20), [utils_1.PrimitiveType.ANY]));
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.function(GlobalFunction.HASH256, new utils_1.BytesType(32), [utils_1.PrimitiveType.ANY]));
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.function(GlobalFunction.CHECKSIG, utils_1.PrimitiveType.BOOL, [utils_1.PrimitiveType.SIG, utils_1.PrimitiveType.PUBKEY]));
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.function(GlobalFunction.CHECKMULTISIG, utils_1.PrimitiveType.BOOL, [new utils_1.ArrayType(utils_1.PrimitiveType.SIG), new utils_1.ArrayType(utils_1.PrimitiveType.PUBKEY)]));
exports.GLOBAL_SYMBOL_TABLE.set(SymbolTable_js_1.Symbol.function(GlobalFunction.CHECKDATASIG, utils_1.PrimitiveType.BOOL, [utils_1.PrimitiveType.DATASIG, new utils_1.BytesType(), utils_1.PrimitiveType.PUBKEY]));
