// cashc/dist/main/artifact/Artifact.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateArtifact = void 0;
const utils_1 = require("@cashscript/utils");
const index_js_1 = require("../index.js");
function generateArtifact(ast, script, source) {
    const { contract } = ast;
    const constructorInputs = contract.parameters
        .map((parameter) => ({ name: parameter.name, type: parameter.type.toString() }));
    const abi = contract.functions.map((func) => ({
        name: func.name,
        inputs: func.parameters.map((parameter) => ({
            name: parameter.name,
            type: parameter.type.toString(),
        })),
    }));
    const bytecode = (0, utils_1.scriptToAsm)(script);
    return {
        contractName: contract.name,
        constructorInputs,
        abi,
        bytecode,
        source,
        compiler: {
            name: 'cashc',
            version: index_js_1.version,
        },
        updatedAt: new Date().toISOString(),
    };
}
exports.generateArtifact = generateArtifact;
